Cuda Warnings: Add empty function to silence warning of unused vars

ndellingwood opened the pull request:
Addresses issue #1767
Variables used in macros within device functions result in warnings
about unused variables. This commit adds function to silence those
warnings.

ndellingwood commented:
@crtrott @ibaned this isn't a pretty solution, but an option to get rid of lots of warnings about unused `mask` variables passed to macros in device functions. Trying the `(void)mask;` trick did not work.

ibaned commented:
Maybe it would be better to introduce a `KOKKOS_IMPL_CUDA_SYNCWARP_NEEDS_MASK` and then do this?
```cpp
#ifdef KOKKOS_IMPL_CUDA_SYNCWARP_NEEDS_MASK
unsigned mask = width==32?0xffffffff:((1<<width)-1)<<((threadIdx.y*blockDim.x+threadIdx.x)%(32/width))*width;
#endif
```
That way we can actually avoid the mask computation altogether if we're just going to ignore it, and we won't get unused variable warnings.

ndellingwood commented:
@ibaned @crtrott I updated the PR to use the macro to remove the mask if it is not needed, and removed the mask argument to the corresponding _SYNCWARP related macros. Does this look closer to how you'd like this handled?

ibaned commented on core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp:
```diff
@@ -642,13 +682,23 @@ struct CudaReductionsFunctor<FunctorType, ArgTag, false, false> {
+#endif
     }
+#ifdef KOKKOS_IMPL_CUDA_SYNCWARP_NEEDS_MASK
+    fcn_to_silence_unused_var_warnings(mask);
```
This `fcn_to_silence_unused_var_warnings` shouldn't be needed anymore?

ndellingwood replied:
I needed this one call after the rest of the changes to silence the warning here but I'm not sure why; the other spots I commented out I'll clean up.

ibaned commented on core/src/Kokkos_Core_fwd.hpp:
```diff
@@ -76,6 +76,10 @@ constexpr AUTO_t AUTO = Kokkos::AUTO_t();
+template <class Type>
+KOKKOS_INLINE_FUNCTION
+void fcn_to_silence_unused_var_warnings ( Type const& input ) {}
+
```
Shouldn't be needed?

ibaned commented on core/src/impl/Kokkos_Atomic_Generic.hpp:
```diff
@@ -298,8 +311,13 @@ T atomic_oper_fetch( const Oper& op, volatile T * const dest ,
+    done_active = KOKKOS_IMPL_CUDA_BALLOT_MASK(done);
+#endif
   }
+//  fcn_to_silence_unused_var_warnings(mask);
```
Remove commented out code?

ndellingwood replied:
Good catch, I'll clean that.

ibaned commented on core/src/Kokkos_Macros.hpp:
```diff
@@ -521,6 +521,7 @@
 #if defined ( KOKKOS_ENABLE_CUDA )
   #if ( 9000 <= CUDA_VERSION )
   #define KOKKOS_IMPL_CUDA_VERSION_9_WORKAROUND
+  #define KOKKOS_IMPL_CUDA_SYNCWARP_NEEDS_MASK
```
This should only be defined if `__CUDA_ARCH__` is defined as well. That may remove the need for the dummy function altogether.

ndellingwood replied:
I'll add that and check, thanks!

ndellingwood replied:
adding a `__CUDA_ARCH__` check ended up breaking the build, I need to look at this a bit more carefully...

ndellingwood commented:
@ibaned fixed the macro like you suggested, fixed the build, cleaned up commented codes and removed the dummy function. Initial check with Cuda/9.1 + gcc/5.3 on Apollo passes with no 'mask' warnings.

ndellingwood commented:
Quick cuda spot-check on Apollo:

```
Running on machine: apollo
Repository Status:  0e5f5e10ba714c8ee100bc5ca24a4d0a7f143286 Cuda 'mask' warnings - fix macros, cleanup, remove empty fcn


Going to test compilers:  cuda/9.1
Testing compiler cuda/9.1
  Starting job cuda-9.1-Cuda_OpenMP-release
  PASSED cuda-9.1-Cuda_OpenMP-release
#######################################################
PASSED TESTS
#######################################################
cuda-9.1-Cuda_OpenMP-release build_time=303 run_time=204
```

Will test on kokkos-dev to get some Cuda/8.0 coverage.

ndellingwood commented:
kokkos-dev cuda spot-check:

```
[ndellin@kokkos-dev TestAllSandia]$ ../../scripts/testing_scripts/test_all_sandia --spot-check cuda
Running on machine: sems
Repository Status:  a0ef8148a00ae5134d76f11584dcf27daeb0bef9 Cuda Warnings: Add empty function to silence warning of unused vars


Going to test compilers:  cuda/8.0.44
Testing compiler cuda/8.0.44
  Starting job cuda-8.0.44-Cuda_OpenMP-release
  PASSED cuda-8.0.44-Cuda_OpenMP-release
#######################################################
PASSED TESTS
#######################################################
cuda-8.0.44-Cuda_OpenMP-release build_time=565 run_time=554
```

ndellingwood commented:
This only affects Cuda code, but running spot-check with rest of the host compilers for completeness.

ndellingwood commented:
@ibaned finished kokkos-dev spot-check, wanted to make sure the cuda stuff finished before without having to wait for the full check in case something was broken.

```
[ndellin@kokkos-dev TestAllSandia]$ ../../scripts/testing_scripts/test_all_sandia --spot-check gcc intel clang
Running on machine: sems
Repository Status:  a0ef8148a00ae5134d76f11584dcf27daeb0bef9 Cuda Warnings: Add empty function to silence warning of unused vars


Going to test compilers:  gcc/5.3.0 gcc/7.3.0 intel/17.0.1 clang/4.0.1
Testing compiler gcc/5.3.0
  Starting job gcc-5.3.0-OpenMP-release
  PASSED gcc-5.3.0-OpenMP-release
Testing compiler gcc/7.3.0
  Starting job gcc-5.3.0-OpenMP-hwloc-release
  PASSED gcc-5.3.0-OpenMP-hwloc-release
  Starting job gcc-7.3.0-Serial-release
  PASSED gcc-7.3.0-Serial-release
Testing compiler intel/17.0.1
  Starting job gcc-7.3.0-Serial-hwloc-release
  PASSED gcc-7.3.0-Serial-hwloc-release
  Starting job intel-17.0.1-OpenMP-release
  PASSED intel-17.0.1-OpenMP-release
Testing compiler clang/4.0.1
  Starting job intel-17.0.1-OpenMP-hwloc-release
  PASSED intel-17.0.1-OpenMP-hwloc-release
  Starting job clang-4.0.1-Pthread_Serial-release
  PASSED clang-4.0.1-Pthread_Serial-release
  Starting job clang-4.0.1-Pthread_Serial-hwloc-release
  PASSED clang-4.0.1-Pthread_Serial-hwloc-release
#######################################################
PASSED TESTS
#######################################################
clang-4.0.1-Pthread_Serial-hwloc-release build_time=198 run_time=181
clang-4.0.1-Pthread_Serial-release build_time=195 run_time=396
gcc-5.3.0-OpenMP-hwloc-release build_time=178 run_time=95
gcc-5.3.0-OpenMP-release build_time=180 run_time=91
gcc-7.3.0-Serial-hwloc-release build_time=151 run_time=171
gcc-7.3.0-Serial-release build_time=152 run_time=170
intel-17.0.1-OpenMP-hwloc-release build_time=438 run_time=154
intel-17.0.1-OpenMP-release build_time=440 run_time=158
```

ibaned commented:
Looks like we're good, merging!

ndellingwood commented:
@ibaned thanks for help with the macro fixes!
